src/ripple/app/misc/impl/ValidatorSite.cpp

```diff
@@ -382,77 +382,89 @@ ValidatorSite::onSiteFetch(
     detail::response_type&& res,
     std::size_t siteIdx)
 {
-    bool shouldRetry = false;
+    // reset internal state to set stage for next timed fetch attempt
+    auto reset = [this](std::unique_lock<std::mutex>& site_lock, Site& site)
     {
-        std::lock_guard <std::mutex> lock_sites{sites_mutex_};
-        try
+        site.activeResource.reset();
+        // setTimer needs this lock, so we have to release
+        site_lock.unlock();
+        std::lock_guard <std::mutex> lock_state{state_mutex_};
+        fetching_ = false;
+        if (! stopping_)
+            setTimer ();
+        cv_.notify_all();
+    };
+
+    // set error disposition and reset()
+    auto setError = [&reset](
+        std::unique_lock<std::mutex>& site_lock,
+        Site& site, std::string const& msg, bool retry)
+    {
+        site.lastRefreshStatus.emplace (
+            Site::Status{clock_type::now(), ListDisposition::invalid, msg});
+        if (retry)
+            site.nextRefresh = clock_type::now() + ERROR_RETRY_INTERVAL;
+        reset (site_lock, site);
+    };
+
+    std::unique_lock <std::mutex> lock_sites{sites_mutex_};
+    try
+    {
+        if (ec)
         {
-            if (ec)
-            {
-                JLOG (j_.warn()) <<
-                    "Problem retrieving from " <<
-                    sites_[siteIdx].activeResource->uri <<
-                    " " <<
-                    ec.value() <<
-                    ":" <<
-                    ec.message();
-                shouldRetry = true;
-                throw std::runtime_error{"fetch error"};
-            }
+            JLOG (j_.warn()) <<
+                "Problem retrieving from " <<
+                sites_[siteIdx].activeResource->uri <<
+                " " <<
+                ec.value() <<
+                ":" <<
+                ec.message();
+            setError (lock_sites, sites_[siteIdx], "fetch error", true);
```

Comment:
I'm not sold on this particular refactor. I don't believe that exceptions should be used for flow control but the exception here seems pretty reasonable, and consolidating that handling in the exception handling makes sense.

I don't feel particularly strongly here and would be interested in additional feedback (from @seelabs, which you already tagged but also @scottschurr  and @HowardHinnant) but my first inclination that this isn't an improvement.

Reply:
Consider using a single `onError` lambda that's called from both the exception and when `ec` is set. Keep this lambda within the scope of the lock, so it's clear the mutex is held, and put the part that was skipped with the thrown exception inside an `else` block paired with the `if (ec)` block. Something like this: https://github.com/seelabs/rippled/commit/306ee04f6ecef31858aa07ec8a8ffe5fb5db35ff

BTW, I thought the existing code was clear, and using the exception like you were doing was good code. My objection had to do with the how we use exceptions in this codebase. My criteria is I should be able to run my dogfood machine in a debugger so it breaks on any exception. If we're throwing exceptions for common events I can't do that. I'm still not sure how I feel about this `throw`. I'll note I hit it twice since the beta came out, so I would like to remove it. However, if we can't do it in a way we all agree is clean I'll modify my dogfood setup so it can deal with exceptions being more common.

Reply:
I like the code the way @mellery451 has it.  However if it _is_ changed, I _strongly_ prefer explicit captures in lambdas such as `auto setError = [&reset]...` as opposed to implicit captures like `auto onError = [&]...`.  It is good to know exactly what is being captured, and what isn't, and the implicit capture makes it to hard to read what is and isn't captured.  The implicit capture makes it too easy to capture something accidentally.

Reply:
I don't understand why is it good to know what is captured. I'd consider it as useful to know what a local block captures. I'll admit many people I respect like explicit captures. The only reason I've heard have to do the lifetime issues - i.e. "you can change the code to pass the lambda to something outside the scope of the function where we have to worry about lifetime issues". I suspect that is because there are many uses of lambdas. In some of those uses we have to worry about lifetimes. In others - include ones like `onError` - we don't. Why annotate the code with something we don't care about?

Reply:
I also don't want to get side-tracked with the lambda capture issue. While I disagree, I'm fine adding lambda captures. While I don't find it useful, if you do I'm happy to add them.

Reply:
One error I've seen in the past is getting a lambda-local confused with an unintended capture.  I don't recall the exact details but it goes along the lines of a local was intended but its declaration was forgotten, and stomped on a capture by reference instead.  Or maybe it was a capture was intended, but a local declaration was added just by think-o.

Reply:
@HowardHinnant You have the same problem with blocks as well.

Reply:
@scottschurr Which argument in particular? Are you worried that we are going to pass `onError`, a lambda that handles errors local to a function, to another function that will spawn a thread and return? I still haven't heard a good argument agaist an implicit capture for this lambda.

The arguments I've heard are:
1) See Scott Meyers. I assume the argument there is one of lifetimes. There is no realistic scenario where `onError` is passed to another thread. Even it it were, any lambda passed to another thread should have explicit captures and we'd change it then.

2) Howard's argument about an unintended capture that was meant to be local to the lambda. We deal with that with local scopes all the time. We could easily rewrite all local scopes to have explict captures, i.e.:
```
foo(){
   int var=0;
   [&var]{
      int var2 = var+1;
   ]()
```
Why capture `var` above?

I like to argue about specific cases. Is there a specific danger you see in the default capture in `onError`?

Reply:
So far both Howard and Scott have liked Mike's refactor over mine. Let me defend why I like mine better:

1) Move the lambda into the scope of the `unique_lock`. This has two effects.
   a) The lambdas no longer need to know about the lock (with an additional change described later). They can assume the mutex is locked.
   b) The unique lock can be changed to a `scoped_lock`, which is easier to reason about. It's locked until the end of the block.

2) Get rid of the early return on line 423 and use an `else` statement instead. This allows us to remove the `reset` lambda and control will flow naturally to the end of the function where the code that's now in `reset` can run naturally. This also means we no longer need to lock and unlock the mutex within `reset` You also no longer need to pass `retry` to the lambdas. The only parameter the lambdas need is the error string. So the refactor takes an interface that looks like this:
```
    auto setError = [&reset](
        std::unique_lock<std::mutex>& site_lock,
        Site& site, std::string const& msg, bool retry)
```
To one that looks like this:
```
        auto onError = [&](std::string const& errMsg)
```

3) The refactor has a much more linear code flow. It calls back into one lambda, and only for error cases. The original calls two lambdas, one witch calls the other, and they are called for both normal and regular cases.

What do you and Howard like better about the origional over mine?

Reply:
@seelabs My call on @mellery451's refactor was not in comparison to yours.  It was simply that @mellery451's refactor seemed good and sufficient to me.  I was not trying to compare yours to his.  Yours may, in fact, be better.  But I think his is a good and sufficient improvement.

Regarding lambda captures, I hear your desire to address specific cases.  And I've not looked hard at this specific case to argue _specifically_ where explicit captures are superior.  So they may not be in this case.

Personally, I like to develop heuristics that will help me write under ordinary circumstances what I feel is the best code I can.  And one of the considerations, for me, is what will happen when the code is maintained.  I prefer code that calls attention to itself when its environment changes.  That's my motivation for preferring, in the general case, explicit captures.  It's also (part of) why I lean toward non-`auto` variable declarations combined with non-equals-sign brace initialization (which prevents narrowing).  It's really just my attempt to improve maintainability of the code.

My hope is to not just have the right code now, I want it to be easy to find what is affected when the code is maintained in the future.  The easiest way I know of is for that code to stop compiling.

Your milage may vary.  These are just my personal heuristics.

Reply:
just wanted to mention my comment on the second refactor:  https://github.com/seelabs/rippled/commit/5c908ffa6bb273d524fbb17410e5dcf388b1a349#r31138632 , although that might be a weak argument about future code being added.